clippy_lints/src/types/redundant_allocation.rs

```diff
@@ -110,3 +109,20 @@ pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_
     }
     true
 }
+
+/// Returns `true` if the allocation would make `hir_ty` go from fat to thin.
+fn alloc_makes_pointer_thin(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>) -> bool {
+    match &hir_ty.kind {
+        TyKind::TraitObject(..) => true,
+        TyKind::Path(ty_qpath) => {
+            let ty_res = cx.qpath_res(ty_qpath, hir_ty.hir_id);
+            if let def::Res::PrimTy(prim_ty) = ty_res {
+                if matches!(prim_ty, PrimTy::Str) {
+                    return true;
+                }
+            }
+            false
+        },
+        _ => false,
+    }
+}
```

Comment:
Since `redundant_allocation::check` isn't called on locals, I think that means it'd never be in the context of a body, which would make it safe to use

```rust
!hir_ty_to_ty(cx.tcx, hir_ty).is_sized(cx.tcx.at(hir_ty.span), cx.param_env)
```

Reply:
Awesome, nice and easy! Moved it inline too and added some tests. I couldn't come up with anything to break `hir_ty_to_ty` in this instance, personally.